rclpy/rclpy/executors.py

```diff
@@ -745,24 +745,45 @@ class SingleThreadedExecutor(Executor):
     def __init__(self, *, context: Optional[Context] = None) -> None:
         super().__init__(context=context)
 
-    def spin_once(self, timeout_sec: Optional[float] = None) -> None:
+        self._futures = []
+
+    def _spin_once_impl(
+        self,
+        timeout_sec: Optional[float] = None,
+        wait_condition: Callable[[], bool] = lambda: False
+    ) -> None:
         try:
-            handler, entity, node = self.wait_for_ready_callbacks(timeout_sec=timeout_sec)
+            handler, entity, node = self.wait_for_ready_callbacks(
+                timeout_sec, None, wait_condition)
         except ShutdownException:
             pass
         except TimeoutException:
             pass
+        except ConditionReachedException:
+            pass
         else:
             handler()
             if handler.exception() is not None:
                 raise handler.exception()
 
+            self._futures.append(handler)
+            # make a copy of the list that we iterate over while modifying it
+            # (https://stackoverflow.com/q/1207406/3753684)
+            for future in self._futures[:]:
+                if future.done():
+                    self._futures.remove(future)
+                    future.result()  # re-raise any exceptions
+
+    def spin_once(self, timeout_sec: Optional[float] = None) -> None:
+        self._spin_once_impl(timeout_sec)
```

Comment:
I'm not sure I totally understand the question, but let me explain what I understand happens to the futures in both `spin_once` and `spin_once_until_future_complete`.

When you are calling `spin_once`, then what happens is that our `wait_condition` for `_spin_once_impl` is the default `lambda: False`.  That means that that condition will never cause the work loop (in `_wait_for_ready_callbacks`) to return early, and so that function will always find one piece of work (assuming it doesn't time out or shutdown).  When it does find that piece of work, it executes it immediately in `_spin_once_impl`.  After that we add it to the `_futures` list to check on whether it is done.  We then *also* look at any other futures that were pending, and if those are done we execute the done handlers for those and remove them from the list.

When you are calling `spin_once_until_future_complete`, the logic is largely the same except that our `wait_condition` is the done callback on the future.  So if the future finishes before we find other work to do, that can cause `wait_for_ready_callbacks` to return early.  If that is the case, then it actually raises `ConditionReachedException`, because we have already called the done callback.  And thus it is never added to `_futures`.

Does that make sense to you?  And does that answer your question at all?